turbopack/crates/turbo-tasks/src/registry.rs

```diff
@@ -1,161 +1,186 @@
+
+        let mut item_to_id = FxHashMap::with_capacity_and_hasher(items.len(), Default::default());
+        let mut names = FxHashSet::with_capacity_and_hasher(items.len(), Default::default());
+
+        let mut id = NonZeroU16::MIN;
+        for &item in items.iter() {
+            item_to_id.insert(item, id.into());
+            let global_name = item.global_name();
+            assert!(
+                names.insert(global_name),
+                "multiple {ty} items registered with name: {global_name}!",
+                ty = T::TYPE_NAME
+            );
+            id = id.checked_add(1).expect("overflowing item ids");
+        }
+
+        Self {
+            id_to_item: items.into_boxed_slice(),
+            item_to_id,
+        }
+    }
+
+    /// Get an item by its ID
+    fn get_item(&self, id: T::Id) -> &'static T {
+        self.id_to_item[*id as usize - 1]
+    }
+
+    /// Get the ID for an item
+    fn get_id(&self, item: &'static T) -> T::Id {
+        match self.item_to_id.get(&item) {
+            Some(id) => *id,
+            None => panic!(
+                "{ty} isn't registered: {item}",
+                ty = T::TYPE_NAME,
+                item = item.global_name()
+            ),
+        }
+    }
+
+    /// Validate that an ID is within the valid range
+    fn validate_id(&self, id: T::Id) -> Option<Error> {
+        let len = self.id_to_item.len();
+        if *id as usize <= len {
+            None
+        } else {
+            Some(anyhow::anyhow!(
+                "Invalid {ty} id, {id} expected a value <= {len}",
+                ty = T::TYPE_NAME
+            ))
+        }
     }
 }
 
-struct Values {
-    id_to_value: Box<[&'static ValueType]>,
-    value_to_id: FxHashMap<&'static ValueType, ValueTypeId>,
-    global_name_to_value: FxHashMap<&'static str, (ValueTypeId, &'static ValueType)>,
+static FUNCTIONS: Lazy<Registry<NativeFunction>> = Lazy::new(|| {
+    let functions = inventory::iter::<CollectableFunction>
+        .into_iter()
+        .map(|c| &**c.0)
+        .collect::<Vec<_>>();
+    Registry::new_from_items(functions)
+});
+
+pub fn get_native_function(id: FunctionId) -> &'static NativeFunction {
+    FUNCTIONS.get_item(id)
+}
+
+pub fn get_function_id(func: &'static NativeFunction) -> FunctionId {
+    FUNCTIONS.get_id(func)
 }
 
-static VALUES: Lazy<Values> = Lazy::new(|| {
+pub fn validate_function_id(id: FunctionId) -> Option<Error> {
+    FUNCTIONS.validate_id(id)
+}
+
+static VALUES: Lazy<Registry<ValueType>> = Lazy::new(|| {
     // Inventory does not guarantee an order. So we sort by the global name to get a stable order
     // This ensures that assigned ids are also stable.
     // We don't currently take advantage of this but we could in the future.  The remaining issue is
     // ensuring the set of values is the same across runs.
-    let mut all_values = inventory::iter::<CollectableValueType>
+    let all_values = inventory::iter::<CollectableValueType>
```

Comment:
```suggestion
    // Inventory does not guarantee an order. So we sort by the global name to get a stable order
    // This ensures that assigned ids are also stable which is important since they are serialized.
    let all_values = inventory::iter::<CollectableValueType>
```